source/material_model/interface.cc

```diff
@@ -306,11 +306,11 @@ namespace aspect
       velocity(fe_values.n_quadrature_points, numbers::signaling_nan<Tensor<1,dim> >()),
       composition(fe_values.n_quadrature_points, std::vector<double>(introspection.n_compositional_fields, numbers::signaling_nan<double>())),
       strain_rate(fe_values.n_quadrature_points, numbers::signaling_nan<SymmetricTensor<2,dim> >()),
-      cell(cell_x == NULL ? NULL : &current_cell),
-      current_cell(cell_x == NULL ? typename DoFHandler<dim>::active_cell_iterator() : *cell_x)
+      cell(cell_x.state() == IteratorState::valid ? &current_cell : NULL),
+      current_cell(cell_x.state() == IteratorState::valid ? cell_x : typename DoFHandler<dim>::active_cell_iterator())
```

Comment:
you can safely just copy `cell_x` here now.

Reply:
For both cell and current_cell or only one of them? I don't really understand why I can just copy it now...

Reply:
Only for the second one.

It's not obvious maybe, but
```
cell_x.state() == IteratorState::valid ? cell_x : typename DoFHandler<dim>::active_cell_iterator()
```
is equivalent to
```
cell_x
```
because
```
 ! (cell_x.state() == IteratorState::valid)
```
implies
```
  cell_x == typename DoFHandler<dim>::active_cell_iterator()
```

Reply:
But we can not copy `active_cell_iterator()` because its copy constructor throws an assertion. That is what confused @jdannberg. This may be fixed in deal.II development, but not 8.5.

Reply:
Oh, doh, yes. That's the annoying difference between theory and practice.

Someone is going to come across this at one point and wonder. Can you write this as
```
#if DEAL_II_VERSION_GTE(9,0)
  current_cell (cell_x)
#else
  current_cell(cell_x.state() == IteratorState::valid ? cell_x : typename DoFHandler<dim>::active_cell_iterator())
#endif
```
to remind ourselves that we can simplify this at a later time?

Reply:
I think it needs to be ``DEAL_II_VERSION_GTE(9,0,0)``